app/javascript/channels/program_channel.ts

```diff
@@ -1,29 +1,58 @@
+import { Program, MaxInputMode } from "../components/types";
 import consumer from "./consumer";
 
-export default (program_id, setProgram) => {
+export interface ProgramChannel {
+  message: (msg: string) => void;
+  setMode: (mode: "anarchy" | "democracy") => void;
+  setMaxInputMode: (mode: MaxInputMode) => void;
+  setVoteInterval: (interval: number) => void;
+  pause: () => void;
+  resume: () => void;
+  resetTick: () => void;
+  clear: () => void;
+  reset: () => void;
+}
+
+export const ProgramChannel = (program: Program, setProgram: (program: Program) => void): ProgramChannel => {
   const sub = consumer.subscriptions.create(
-    { channel: "ProgramChannel", id: program_id },
+    { channel: "ProgramChannel", id: program.id },
     {
       connected() {
-        console.log("Connected to Program Channel(" + program_id + ")"); // Called when the subscription is ready for use on the server
+        console.debug(`Connected to Program Channel(${program.id})`);
       },
 
       disconnected() {
-        console.log("disconnected"); // Called when the subscription has been terminated by the server
+        console.debug("disconnected");
+        setProgram({} as Program);
       },
 
-      received(data) {
-        console.log(`Data:`, data); // Called when there's incoming data on the websocket for this channel
-        setProgram(data);
+      received({ action, data }) {
+        switch (action) {
+          case "tick":
+            console.debug("tick");
+            setProgram({ ...program, ...data });
+            break;
+
+          default:
+            console.log(`Data:`, data);
+            setProgram(data);
+            break;
+        }
       },
     }
   );
   return {
-    message: (msg) => {
-      sub.perform("message", msg);
-    },
-    clear: () => {
+    message: (msg: string) => sub.perform("message", msg),
+    setMode: (mode: "anarchy" | "democracy") => sub.perform("set_mode", { data: mode }),
+    setMaxInputMode: (mode: MaxInputMode) => sub.perform("set_max_input_mode", { data: mode }),
+    setVoteInterval: (interval: number) => sub.perform("set_vote_interval", { data: interval }),
+    pause: () => sub.perform("pause"),
+    resume: () => sub.perform("resume"),
+    resetTick: () => sub.perform("reset_tick"),
+    clear: () => sub.perform("clear"),
+    reset: () => {
       sub.perform("clear");
-    }
+      sub.perform("reset_tick");
```

Comment:
I wish this didn't *look* so messy, but it is nice having a single place where *all* messages to and from the web socket have to go through, makes it very clear where to look when you need to investigate something.